nbsafety/safety.py

```diff
@@ -394,6 +394,54 @@ def _resync_symbols(self, symbols: Set[DataSymbol]):
                 del self.namespaces[dsym.obj_id]
                 self.namespaces[namespace.obj_id] = namespace
             dsym.update_obj_ref(obj)
+  
+    def get_dependencies(self, cell_num: int) -> dict:
+        """
+        Gets a dictionary object of cell dependencies for the last or 
+        currently executed cell.
+
+        Args:
+            - cell_num (int): cell to get dependencies for, defaults to last
+                execution counter
+        
+        Returns:
+            - dict (int, str): map from required cell number to code
+                representing dependencies
+        """
+        dependencies = set()
+        self._get_dependencies(cell_num, dependencies)
+        return {num: self.cell_content_by_counter[num] for num in dependencies}
+    
+    def _get_dependencies(self, cell_num: int, dependencies: set) -> None:
+        """
+        For a given cell, this function recursively populates a set of
+        cell numbers that the given cell depends on, based on the live symbols.
+
+        Args:
+            - dependencies (set<int>): set of cell numbers so far that exist
+            - cell_num (int): current cell to get dependencies for
+        
+        Returns:
+            None
+        """
+        # Base case: cell already in dependencies
+        if cell_num in dependencies:
+            return
+
+        # Add current cell to dependencies
+        dependencies.add(cell_num)
+
+        # Retrieve cell numbers for the dependent symbols
+        cell = self.cell_content_by_counter[cell_num]
+        symbols = self._check_cell_and_resolve_symbols(cell)
+        live_symbols = symbols['live']
+        dep_cell_nums = set([dep_symbol.defined_cell_num for dep_symbol in live_symbols])
```

Comment:
In addition to adding defined_cell_num's of live_symbols to `dep_cell_nums`, we may also want to add defined_cell_nums of those symbols for which `last_used_cell_num` is `cell_num`. The liveness checker is pretty good at traversing function bodies, but will struggle to detect highly dynamic usages of symbols. For example:

```
# cell 1
x = 5

# cell 2
class Foo:
    def foo(self):
        return x

# cell 3
def foo():
    return Foo()
```

If I then have the following cell:

```
print(foo().foo())
```

The current liveness checker will only detect that we need cells 2 and 3 as dependencies -- but we also need cell 1.

Reply:
Conversely, there are cases where liveness checking, which is more conservative, may give us more symbols than what we would get from `last_used_cell_num`. Example:
```
if True:
    print(x)
else:
    print(y)
```
Probably we should get the cells necessary for both `x` and `y` if we want the program necessary to reconstruct the above cell's output -- the liveness checker will ensure we get both, while relying on the dynamic last usage via `last_defined_cell_num` would leave out the cell needed for generating `y`.

Reply:
In summary, we probably want both.